### ISSUE TITLE
feature request: Extended OAuth2 + LDAP configuration options

### ISSUE THREAD

**Issue by nwrkbiz** [docker-mailserver/docker-mailserver#3830]
### Context

We did set up a rather complex DMS + openldap + OAuth2 (keycloak) stack and found some configurations that would be benifitial to expose via environment variables. Namely:

- /etc/dovecot/dovecot-ldap.conf.ext -> tls_require_cert (allows using self signed certificates ond/or setting the verification mode)
- /etc/dovecot/conf.d/10-auth.conf -> auth_username_format
- /etc/dovecot/dovecot-oauth2.conf.ext -> username_attribute

### Description

Allow setting the configs described above using following env variables:

OAUTH2_USERNAME_ATTRIBUTE
LDAP_TLS_REQUIRE_CERT
DOVECOT_AUTH_USERNAME_FORMAT

### Alternatives

We now did hack some echo/sed magic into the docker-compose command attribute.

### Applicable Users

DMS-Administrator

### Are you going to implement it?

Yes, because I know the probability of someone else doing it is low and I can learn from it.

### What are you going to contribute?

Not sure how fast I can provide these changes, i somebody hints me to the files I need to touch I would be glad.

**Comment by georglauterbach** [comment]
I assigned the maintainer that has the most experience with OAUTH2. I will unsubscribe from this issue, though (too many balls in the air); ping me if you need me again.

**Comment by polarathene** [comment]
> * `/etc/dovecot/dovecot-ldap.conf.ext` -> `tls_require_cert` (allows using self signed certificates and/or setting the verification mode)

This is possible once I get [this PR merged](https://github.com/docker-mailserver/docker-mailserver/pulls) (_pending on docs, which requires me to not be spending time elsewhere in DMS_), although it's not likely to make it for v14 and as it's a breaking change will be deferred to v15.

> * `/etc/dovecot/conf.d/10-auth.conf` -> `auth_username_format` (keycloak is configured to just provide the username and not a full e-mail address, we configured this to %n)

Are you sure that one is required? I don't recall it being required for LDAP which is what this context is really related to IIRC.

Just looked up this setting and `%n` drops the domain only when it's given. If you're still getting the domain despite omitting it, that's due to something else which I'm aware of as an issue with LDAP and is on my TODO. As such I don't consider this a required fix as it's not the source of the issue.

> * `/etc/dovecot/dovecot-oauth2.conf.ext` -> `username_attribute` (keycloak is configured to just provide the username and not a full e-mail address, we configured this to preferred_username)

The OAuth2 feature got merged early to avoid being stale, I wasn't a fan of the final config approach but stalling the merge may have been problematic. Plans are to adopt the same config templates feature instead.

I am reluctant towards ENV configs as they're unnecessary, you can provide these configs via volume mount files or compose `configs.content` feature if you must inline it within the same `compose.yml`, it's only a little bit more verbose (especially when multiple ENV would otherwise be required).

The config templates feature will simplify that by allowing more flexibility (_while still supporting ENV override implicitly_).

---

> We now did hack some echo/sed magic into the docker-compose command attribute, thats applies the settings described above right before starting DMS.

You should be handling this via [`user-patches.sh`](https://docker-mailserver.github.io/docker-mailserver/v13.3/config/advanced/override-defaults/user-patches/). It'll apply the changes after setup scripts have run and configured everything else, but just before services are actually started.

You'll need to wait on the above concerns being addressed in a future release before you can drop the `user-patches.sh` workaround.

These are ENV requests that I'm personally against adding into DMS (_an ENV that serves the single purpose of updating one setting in a config file, with no real logic beyond applying the value_).

**Comment by nwrkbiz** [comment]
> This is possible once I get [this PR merged](https://github.com/docker-mailserver/docker-mailserver/pulls) (pending on docs, which requires me to not be spending time elsewhere in DMS), although it's not likely to make it for v14 and as it's a breaking change will be deferred to v15.

O.k., nice!

> Are you sure that one is required? I don't recall it being required for LDAP which is what this context is really related to IIRC.
Might be that we do not need this anymore since we changed the username attribute to cn using (I would need to verify this tho):
`DOVECOT_USER_ATTRS=cn=user,homeDirectory=home,=uid=5000,=gid=5000`

But before we applied this change, the imap login failed without %n.

**Comment by polarathene** [comment]
> But before we applied this change, the imap login failed without %n.

Presently the LDAP tests have this configuration and notes:

https://github.com/docker-mailserver/docker-mailserver/blob/47f8d50bebaa0b1f1cb85d45c1d9210d9ce80a49/test/tests/serial/mail_with_ldap.bats#L85-L90

Also this if using `ENABLE_SASLAUTHD=1` (which you don't actually need, but I mention it because many have it enabled):

https://github.com/docker-mailserver/docker-mailserver/blob/47f8d50bebaa0b1f1cb85d45c1d9210d9ce80a49/test/tests/serial/mail_with_ldap.bats#L74-L83

**Comment by nwrkbiz** [comment]
> Also this if using `ENABLE_SASLAUTHD=1` (which you don't actually need, but I mention it because many have it enabled):
> 
> https://github.com/docker-mailserver/docker-mailserver/blob/47f8d50bebaa0b1f1cb85d45c1d9210d9ce80a49/test/tests/serial/mail_with_ldap.bats#L74-L83

Hmm, but in postfix the query match is not the attribute passed to the backend. The query is just used to match an entry. The matched entries attribute to be passed is defined by a different setting which is hardcoded to `mail` as far what we have observed.

**Comment by polarathene** [comment]
> The matched entries attribute to be passed is defined by a different setting (`result_attribute`) which is hardcoded to `mail` for all postfix ldap settings as far as we have observed.

Yes, it is presently:

https://github.com/docker-mailserver/docker-mailserver/blob/47f8d50bebaa0b1f1cb85d45c1d9210d9ce80a49/target/postfix/ldap-users.cf#L5

Also the default query filter for saslauthd:

https://github.com/docker-mailserver/docker-mailserver/blob/47f8d50bebaa0b1f1cb85d45c1d9210d9ce80a49/target/scripts/startup/setup.d/saslauthd.sh#L20

My docs rewrite for LDAP does touch on that and it'll be easy to configure once the new approach is merged.

---

If SASLAUTHD is enabled:
- The domain-part is stripped away (_which is inconvenient if you have multiple domains/realms_), so `john.doe@example.com` and `john.doe@elsewhere.com` both send a user query to LDAP for `john.doe` (_regardless of providing `%u` instead of `%U`, since `ldap` is not configured with `-r` like `rimap` mechanism is for SASLAUTHD_).
- The value returned by `result_attribute` would then be compared to the login username that Postfix has AFAIK (_typically `user@domain`_), which at this point appends `@$myorigin` when no domain-part was given in the login username.

That caveat isn't an issue when using Dovecot as the default SASL provider, so better to not use the saslauthd feature when it's not serving you any value (_it was contributed for when Dovecot was not used_).

`result_attribute = mail` is relevant to Postfix not for login IIRC, but for other lookups such as what users exist.

---

> When using dovecot directly this is not a problem, as we can work this around using %u ("just use the localpart") in the dovecot query.

You should have a similar experience with Postfix if you disable SASLAUTHD, authentication then becomes Postfix => Dovecot => LDAP, instead of Postfix => SASLAUTHD => LDAP.

**Comment by NhProGamer** [comment]
so, what is the correct solution for making that working correctly ?

**Comment by polarathene** [comment]
> correct solution for making that working correctly

Be more specific? What in the above discussion are you referring to?

The OAuth feature was merged and released earlier than desired, the improved config support will depend upon the LDAP improvements PR to be merged first (will not arrive in v14, must wait until v15 release of DMS).

You can manually manage your config changes with `user-patches.sh` in the meantime.
